pkg/updater/repository/repository.go

```diff
@@ -0,0 +1,324 @@
+// 3. Cleanup the repository to remove the previous experiment package.
+func (r *Repository) DeleteExperiment() error {
+	repository, err := openRepository(r.RootPath)
+	if err != nil {
+		return err
+	}
+	err = repository.cleanup()
+	if err != nil {
+		return fmt.Errorf("could not cleanup repository: %w", err)
+	}
+	if !repository.stable.Exists() {
+		return fmt.Errorf("stable package does not exist, invalid state")
+	}
+	if !repository.experiment.Exists() {
+		return nil
+	}
+	err = repository.experiment.Delete()
+	if err != nil {
+		return fmt.Errorf("could not delete experiment link: %w", err)
+	}
+	err = repository.cleanup()
+	if err != nil {
+		return fmt.Errorf("could not cleanup repository: %w", err)
+	}
+	return nil
+}
+
+type repository struct {
+	rootPath string
+
+	stable     *link
+	experiment *link
+}
+
+func openRepository(rootPath string) (*repository, error) {
+	stableLink, err := newLink(filepath.Join(rootPath, stableVersionLink))
+	if err != nil {
+		return nil, fmt.Errorf("could not load stable link: %w", err)
+	}
+	experimentLink, err := newLink(filepath.Join(rootPath, experimentVersionLink))
+	if err != nil {
+		return nil, fmt.Errorf("could not load experiment link: %w", err)
+	}
+
+	return &repository{
+		rootPath:   rootPath,
+		stable:     stableLink,
+		experiment: experimentLink,
+	}, nil
+}
+
+func (r *repository) setExperiment(name string, sourcePath string) error {
+	path, err := movePackageFromSource(name, r.rootPath, sourcePath)
+	if err != nil {
+		return fmt.Errorf("could not move experiment source: %w", err)
+	}
+	return r.experiment.Set(path)
+}
+
+func (r *repository) setStable(name string, sourcePath string) error {
+	path, err := movePackageFromSource(name, r.rootPath, sourcePath)
+	if err != nil {
+		return fmt.Errorf("could not move stable source: %w", err)
+	}
+	return r.stable.Set(path)
+}
+
+func movePackageFromSource(packageName string, rootPath string, sourcePath string) (string, error) {
+	if packageName == "" || packageName == stableVersionLink || packageName == experimentVersionLink {
+		return "", fmt.Errorf("invalid package name")
+	}
+	targetPath := filepath.Join(rootPath, packageName)
+	_, err := os.Stat(targetPath)
+	if err == nil {
+		return "", fmt.Errorf("target package already exists")
+	}
+	if !errors.Is(err, os.ErrNotExist) {
+		return "", fmt.Errorf("could not stat target package: %w", err)
+	}
+	err = os.Rename(sourcePath, targetPath)
+	if err != nil {
+		return "", fmt.Errorf("could not move source: %w", err)
+	}
+	return targetPath, nil
+}
+
+func (r *repository) cleanup() error {
+	files, err := os.ReadDir(r.rootPath)
+	if err != nil {
+		return fmt.Errorf("could not read root directory: %w", err)
+	}
+	for _, file := range files {
+		isLink := file.Name() == stableVersionLink || file.Name() == experimentVersionLink
+		isStable := r.stable.Exists() && r.stable.Target() == file.Name()
+		isExperiment := r.experiment.Exists() && r.experiment.Target() == file.Name()
```

Comment:
We access `file.Name()` multiple times. Should we store that in a local variable?

Reply:
Could we optimize this checks? Could a single file be called `isLink`, `isStable` and `isExperiment`? If they are exclusive, we could probably change the code so if  a file is `isLink` we do not compute `isStable` and `isExperiment`

Reply:
It's just a bunch of string compare so it's going to be extremely cheap compared to the IO to list the files for example. I'd rather keep the code a bit more readable if you don't feel strongly about it. 

`.File()` is just a getter: https://cs.opensource.google/go/go/+/refs/tags/go1.21.6:src/os/file.go;l=56, I don't see a big win introducing a variable either. Once compiled it's likely all the same anyway so I'd only see it justified if it provided a big readability win